src/mcpwm/timer.rs

```diff
@@ -0,0 +1,302 @@
+    ///         v                    .                    v|
+    ///         .                    . <----  duty  ----> .v
+    ///         .                    .--------------------..
+    ///         .       Active       |                    |.
+    ///         .                    |                    |.
+    ///         .     Not active     |      Active        |.
+    ///         ----------------------                    ----
+    /// ```
+    Down,
+
+    /// Symmetric mode. The counter will start from its lowest value and increment every clock cycle until the period is reached
+    ///
+    /// The wave form will end up looking something like the following:
+    /// ```
+    ///                                             change count dir to decrement, counter = period
+    ///       start, counter = 0, incrementing          |                                     change count dir to increment, counter = 0
+    ///         |                                       |                                        |
+    ///         |                                       |*--- counter = period                   |*----- start, counter = 0, incrementing
+    ///         v <----  duty  ----> .                  v|                  . <----  duty  ----> ||
+    ///         .                    .                  .v                  .                    vv
+    ///         ---------------------.                  ..                  .-------------------------------------------.                  ..                  .--
+    ///                 Active       |                  ..                  |        Active                Active       |                  ..                  |
+    ///                              |                  ..                  |                                           |                  ..                  |
+    ///                              |     Not active   ..    Not active    |                                           |     Not active   ..    Not active    |
+    ///                              ----------------------------------------                                           ----------------------------------------
+    /// ```
+    /// NOTE: That in this mode, the frequency will be half of that specified
+    UpDown,
+
+    /// Timer paused
+    Pause,
+}
+
+impl From<CountMode> for mcpwm_timer_count_mode_t {
+    fn from(val: CountMode) -> Self {
+        match val {
+            //CounterMode::Frozen => mcpwm_counter_type_t_MCPWM_FREEZE_COUNTER,
+            CountMode::Up => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP,
+            CountMode::Down => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_DOWN,
+            CountMode::UpDown => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP_DOWN,
+            CountMode::Pause => mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_PAUSE,
+        }
+    }
+}
+
+pub struct TIMER<const N: u8, G: Group> {
```

Comment:
I've asked for the removal of as many traits from this impl as possible, but introducing a `Timer` trait here, which is a marker trait for a `TIMER` **peripheral** (don't confuse it with the existing `Timer` struct which should be renamed to `TimerDriver` - see below)), has the big benefit that it will help you to remove a lot of the generics from the driver (`TimerDriver`).